spec/02-integration/02-cmd/02-start_stop_spec.lua

```diff
@@ -39,6 +41,102 @@ describe("kong start/stop", function()
     assert(helpers.kong_exec("start --conf " .. helpers.test_conf_path))
     assert.truthy(helpers.path.exists(helpers.test_conf.kong_env))
   end)
+  it("start without daemonizing will properly log to stdout/stderr", function()
+    --[[
+    What are we testing exactly?
+    When Kong starts, the cli commands will shell out to start the 'nginx'
+    executable. The stdout/stderr of this command is redirected to be captured
+    in case of errors.
+    A problem arises when Kong runs in the foreground 'nginx_daemon=off', and
+    the nginx log output is set to '/dev/stderr' and '/std/stdout'.
+    In this case the shell command to start Kong will capture all output send
+    to stdout/stderr by nginx. Which causes the tempfiles to grow uncontrolable.
+
+    So when Kong is run in the foreground, no redirects should be used when
+    starting Kong to prevent the above from happening. Or in other words we
+    want the stdout/stderr of the "kong start" command to receive all output
+    of the nginx logs. Instead of them being swallowed by the intermediary
+    nginx start command (used by "kong start" under the hood).
+    --]]
+    local xos = _G.os  -- trick the linter here in accepting global updates
+    local exec = xos.execute
+    local stdout = pl_path.tmpname()
+    local stderr = pl_path.tmpname()
+    -- create a finalizer that will stop Kong and cleanup logfiles
+    finally(function()
+      os.remove(stdout)
+      os.remove(stderr)
+      xos.execute = exec
+      assert(helpers.kong_exec("stop", {
+        prefix = helpers.test_conf.prefix,
+      }))
+    end)
+    -- catch the prepared start command (let the 'helpers' do
+    -- the hard work of building the command)
+    local start_cmd
+    xos.execute = function(cmd) start_cmd = cmd return true end
+    assert(helpers.kong_exec("start", {
+      prefix = helpers.test_conf.prefix,
+      database = helpers.test_conf.database,
+      pg_database = helpers.test_conf.pg_database,
+      cassandra_keyspace = helpers.test_conf.cassandra_keyspace,
+      proxy_listen = "127.0.0.1:" .. helpers.test_conf.proxy_port,
+      admin_listen = "127.0.0.1:" .. helpers.test_conf.admin_port,
+      admin_access_log = "/dev/stdout",
+      admin_error_log = "/dev/stderr",
+      proxy_access_log = "/dev/stdout",
+      proxy_error_log = "/dev/stderr",
+      log_level = "debug",
+      nginx_daemon = "off",
+    }))
+    xos.execute = exec
+    -- remove the stdout/stderr redirects from the captured command
+    -- and insert new ones for us to track
+    start_cmd = start_cmd:match("^(.- bin/kong start).-$")
+    start_cmd = start_cmd .. " > " .. stdout .. " 2> " .. stderr
+    start_cmd = start_cmd .. " &"  -- run it detached
```

Comment:
I wonder if would it make sense to add a feature in `helpers.start_kong` to make it do that, instead of making the test rely on knowledge of its internals.

Reply:
we have that, but it runs Kong as a daemon. And in this case we cannot use that since we run in foreground, and then that call would block, stopping the tests.

Reply:
I meant adding a flag to `start_kong` that adds the redirects and ` &`... but at this point I'm still wondering if it's all worth it. As we discussed earlier, this is probably going to remain a fragile test for a very edge case... Would like to hear more opinions!

Reply:
I think the following test case is a lot more concise and reliable:
```lua
  it("#oo", function()
    local pl_utils = require "pl.utils"
    local pl_file = require "pl.file"

    local stdout_path = os.tmpname()

    finally(function()
      os.remove(stdout_path)
    end)

    local cmd = string.format("KONG_PROXY_ACCESS_LOG=/dev/stdout "    ..
                              "KONG_NGINX_DAEMON=off %s start -c %s " ..
                              ">%s 2>/dev/null &", helpers.bin_path,
                              helpers.test_conf_path, stdout_path)

    local ok, _, _, stderr = pl_utils.executeex(cmd)
    if not ok then
      error(stderr)
    end

    do
      local proxy_client

      -- get a connection, retry until kong starts
      helpers.wait_until(function()
        proxy_client = helpers.proxy_client()
        return true
      end)

      local res = assert(proxy_client:send {
        method = "GET",
        path = "/hello",
      })
      assert.res_status(404, res) -- no API configured
    end

    assert(helpers.stop_kong(helpers.test_conf.prefix))

    -- TEST: since nginx started in the foreground, the 'kong start' command
    -- stdout should receive all of nginx's stdout as well.
    local stdout = pl_file.read(stdout_path)
    assert.matches([["GET /hello HTTP/1.1" 404]] , stdout, nil, true)
  end)
```